### DISCUSSION TITLE
How to determine if a rebase is needed for an MR

### DISCUSSION THREAD

**Discussion by FirefighterBlu3** [python-gitlab/python-gitlab#2006] [Q&A]
After an MR is created, how do we:

1. determine that a (ff) merge is not possible and
2. get the message telling us why?

example messages:

- source branch is X commits behind target
- FF merge not possible, rebase src to target
- FF merge not possible, rebase locally

please note that _mr.merge_status == 'can_be_merged'_ and _mr.has_conflicts == False_

**Comment by JohnVillalovos** [comment]
Has the project been set to require only fast-forward merge? If so then the status should not be `can_be_merged` if it needs to be rebased.

Also there is `mr.diverged_commits_count` which if it is not zero means it needs to be rebased.

I look at the following:

A merge request should be rebased if it meets the following conditions:
  * The target branch is the project's default branch
  * The pipeline succeeded
  * The MR does not have conflicts.
  * It has diverged from the target branch.
  * There is no rebase_in_progress for the MR
  * There is no merge_error
  * It is not a work_in_progress

**Reply by nejch** [reply]
Ah good point. I guess if that's what @FirefighterBlu3 needs then the project should just have that enforced.

**Comment by nejch** [comment]
Hi @FirefighterBlu3, I think you're looking for the `include_diverged_commits_count` attribute in the `get()` call and then look for the `diverged_commits_count` in the response.

Otherwise, this largely depends on whether the information is available from the API. These statuses are calculated by the server asynchronously (https://docs.gitlab.com/ee/api/merge_requests.html#single-merge-request-response-notes), so you'd need to continuously poll to get the right status. I suggest you look through the upstream docs for hints.

If this information is not available from the server, then unfortunately there isn't much python-gitlab can do as it just wraps the API calls.
In that case, I'd maybe try to just `merge()` in a try/except block and see if the error message provided by the server is helpful enough.
